// ifmcommon/dist/common/func/generate.exception.logobject.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createExceptionReqResLogObj = void 0;
const common_1 = require("@nestjs/common");
const exception_type_1 = require("../const/exception.type");
function createExceptionReqResLogObj(request, exception, errorType) {
    let status = exception instanceof common_1.HttpException
        ? exception.getStatus()
        : common_1.HttpStatus.INTERNAL_SERVER_ERROR;
    if (errorType === exception_type_1.ExceptionType.MONGO_EXCEPTİON) {
        status = exception.code;
    }
    const method = request.method;
    const requestInformation = {
        timestamp: new Date(),
        user: request.user || null,
        path: request.url,
        method,
        body: request.body,
    };
    const errorResponseLog = {
        timestamp: new Date().toLocaleDateString(),
        path: request.url,
        method,
        status,
        message: exception.message,
        errorType,
    };
    const reqResObject = { requestInformation, errorResponseLog };
    return reqResObject;
}
exports.createExceptionReqResLogObj = createExceptionReqResLogObj;
